[CMake] Version check fails

### What is the bug?

In our CMake file we have this line:

```cmake
find_package(GDAL 3.3 CONFIG REQUIRED)
```
But this fails in case we building a dev build of CMake

```
CMake Error at CMakeLists.txt:38 (find_package):
  Could not find a configuration file for package "GDAL" that is compatible
  with requested version "3.3".

  The following configuration files were considered but not accepted:

    C:/dev/dependencies/gdal/lib/cmake/gdal/GDALConfig.cmake, version: 3.11.0dev
```

Note: Had this issue already on our CI when checking our via
```sh
git clone  --depth 1 --branch v3.10.0 https://github.com/OSGeo/gdal.git .
```
Shouldn't the `dev` removed on the commits with a release tag? Didn't checked the CMake source in which cases the `dev` bill be added and when not.

### Steps to reproduce the issue

See bug decription

### Versions and provenance

Windows & Linux

### Additional context

_No response_

> > , but nothing is removed and ABI stability is irrelevant when building (
> 
> There are sometimes C++ API breakage in the 3.x series (tend to affect "obscure" parts of the API)

Ah okay, in this case this makes sense. Do you know, how often this happens?

> With #11335 I believe there's nothing left preventing this ticket to be close?

Mhm, I think there is room for improvement
- Version range support would be nice, so we could define a range, where we know there is breaking API change for us, as currently you get:
  ```
  CMake Error at CMakeLists.txt:38 (find_package):
    Could not find a configuration file for package "GDAL" that is compatible
    with requested version range "3.7...3.11".

    The following configuration files were considered but not accepted:

      C:/dev/dependencies/gdal/lib/cmake/gdal/GDALConfig.cmake, version: 3.10.0dev
  ```
- The error message could be improved so that you get a hint that only the same minor version is compatible and the same major version is not sufficient. Currently it is a bit irritating, that `find_package(GDAL 3.7 MODULE REQUIRED)` is fine with GDAL 3.10, but `find_package(GDAL 3.7 CONFIG REQUIRED)` not.

I think both would require changes on the part of CMake. I will write a ticket to them and leave the ticket open here until it is clear whether something could be improved on the part of GDAL or not (in the latter case, the ticket could then be closed).

--- a/doc/source/development/cmake.rst
+++ b/doc/source/development/cmake.rst
@@ -10,7 +10,7 @@ The recommended way to use the GDAL library 3.5 or higher in a CMake project is 
 link to the imported library target ``GDAL::GDAL`` provided by
 the CMake configuration which comes with the library. Typical usage is:
 
-.. code::
+.. code:: cmake
 
     find_package(GDAL CONFIG REQUIRED)
 
@@ -26,10 +26,26 @@ the cache variable or environment variable ``CMAKE_PREFIX_PATH``. In
 particular, CMake will consult (and set) the cache variable
 ``GDAL_DIR``.
 
+If a specific minor version is required, you can search for this via:
+
+.. code:: cmake
+
+    find_package(GDAL 3.10 CONFIG REQUIRED)
+
+If more than one minor version is to be supported at the same time,
+``${GDAL_VERSION}`` itself must be evaluated.
+
+.. code:: cmake
+
+    find_package(GDAL CONFIG REQUIRED)
+    if(GDAL_VERSION VERSION_LESS "3.7" OR GDAL_VERSION VERSION_GREATER "3.9")
+      message(FATAL_ERROR "Required at least GDAL version 3.7 - 3.9, but found ${GDAL_VERSION}.")
+    endif()
+
 Before GDAL 3.5, it is recommended to use `find module supplied with CMake <https://cmake.org/cmake/help/latest/module/FindGDAL.html>`__.
 This also creates the ``GDAL::GDAL`` target. It requires CMake version 3.14.
 
-.. code::
+.. code:: cmake
 
     cmake_minimum_required(VERSION 3.14)
 

